main.go

```diff
@@ -201,39 +226,51 @@ func slurpFiles(paths []string, program string, encoder formats.Encoding, flags
 }
 
 type fileInfo struct {
-	path             string
-	data             []byte
-	encoder, decoder formats.Encoding
+	path   string
+	reader io.Reader
+	data   []byte
+	read   bool
 }
 
-func readFile(path string, flags flags) (*fileInfo, error) {
-	path = os.ExpandEnv(path)
-	fileBytes, err := ioutil.ReadFile(path)
+func (info *fileInfo) MarshalJSONBytes(decoder formats.Encoding) ([]byte, error) {
+	fileBytes, err := info.GetContents()
 	if err != nil {
-		return nil, fmt.Errorf("failed to read file at %s: `%s`", path, err)
-	}
-
-	// If there was no input, there's no output!
-	if len(fileBytes) == 0 {
-		return nil, nil
+		return nil, err
 	}
-
-	decoder, err := determineDecoder(flags.inputFormat, path, fileBytes)
+	data, err := decoder.MarshalJSONBytes(fileBytes)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to jsonify file at %s: `%s`", info.path, err)
 	}
+	return data, err
+}
 
-	encoder, err := determineEncoder(flags.outputFormat, decoder)
-	if err != nil {
-		return nil, err
+func (info *fileInfo) GetContents() ([]byte, error) {
+	if !info.read {
+		var err error
+		info.data, err = ioutil.ReadAll(info.reader)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read file at %s: `%s`", info.path, err)
+		}
+
+		if readCloser, ok := info.reader.(io.ReadCloser); ok {
```

Comment:
We should probably make this check first and defer the `Close()`. If ReadAll fails for some reason, the fd will remain open.

Reply:
Done